Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm

```diff
@@ -1396,6 +1461,160 @@
+{
+    TestWebKitAPI::HTTPServer server({
+        { "/"_s, { { { "Content-Type"_s, "text/html"_s } }, "<iframe src='/frame.html'><html><body onload=\"document.body.style.background ='blue'\"/></html></iframe>"_s } },
+        { "/frame.html"_s, { { { "Content-Type"_s, "text/html"_s } }, ""_s } },
+    }, TestWebKitAPI::HTTPServer::Protocol::Http);
+
+    static auto *manifest = @{
+        @"manifest_version": @2,
+
+        @"name": @"Tabs Test",
+        @"description": @"Tabs Test",
+        @"version": @"1",
+
+        @"background": @{
+            @"scripts": @[ @"background.js" ],
+            @"type": @"module",
+            @"persistent": @NO,
+        },
+    };
+
+    static auto *javaScript = @"document.body.style.background = 'pink'";
+
+    auto *backgroundScript = Util::constructScript(@[
+        @"const tabs = await browser.tabs.query({ active: true, currentWindow: true })",
+        @"const tabId = tabs[0].id",
+
+        @"let results = await browser.tabs.executeScript(tabId, { allFrames: false, file: 'executeScript.js' })",
+        @"browser.test.assertEq(results[0], 'pink')",
+
+        @"results = await browser.tabs.executeScript(tabId, { frameId: 0, file: 'executeScript.js' })",
+        @"browser.test.assertEq(results[0], 'pink')",
+
+        @"results = await browser.tabs.executeScript(tabId, { allFrames: true, file: 'executeScript.js' })",
+        @"browser.test.assertEq(results[0], 'pink')",
+        @"browser.test.assertEq(results[1], 'pink')",
+
+        @"results = await browser.tabs.executeScript(tabId, { allFrames: false, code: \"document.body.style.background = 'pink'\" })",
+        @"browser.test.assertEq(results[0], 'pink')",
+
+        @"browser.test.notifyPass()"
+    ]);
+
+    auto extension = adoptNS([[_WKWebExtension alloc] _initWithManifestDictionary:manifest resources:@{ @"background.js": backgroundScript, @"executeScript.js": javaScript }]);
+    auto manager = adoptNS([[TestWebExtensionManager alloc] initForExtension:extension.get()]);
+
+    auto *urlRequest = server.requestWithLocalhost();
+    auto *url = urlRequest.URL;
+    auto *matchPattern = [_WKWebExtensionMatchPattern matchPatternWithScheme:url.scheme host:url.host path:@"/*"];
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forMatchPattern:matchPattern];
+    [manager.get().defaultTab.mainWebView loadRequest:urlRequest];
+
+    [manager loadAndRun];
+
+    // Manifest v3 deprecates executeScript(), so it should be an undefined property.
+
+    backgroundScript = Util::constructScript(@[
+        @"browser.test.assertEq(typeof browser.tabs.executeScript, 'undefined')",
+        @"browser.test.assertEq(browser.tabs.executeScript, undefined)",
+
+        // Finish
+        @"browser.test.notifyPass()"
+    ]);
+
+    Util::loadAndRunExtension(tabsManifest, @{ @"background.js": backgroundScript });
+}
+
+TEST(WKWebExtensionAPITabs, InsertAndRemoveCSS)
```

Comment:
Combined insertCSS and removeCSS into one test since you need to call insertCSS in order to remove it with removeCSS